### DISCUSSION TITLE
Getting 407 when using proxy

### DISCUSSION THREAD

**Discussion by aguilardias** [quarkiverse/quarkus-cxf#941] [Q&A] [answered]
i have to use a proxy with username and password, to call a SOAP webservice.

im using this
```
@PostConstruct
void configurePortType() throws NoSuchAlgorithmException, UnrecoverableKeyException, KeyStoreException {
	final HTTPConduit httpConduit = (HTTPConduit) ClientProxy.getClient(someSoap).getConduit();
        HTTPClientPolicy httpClientPolicy = new HTTPClientPolicy();
        httpClientPolicy.setAllowChunking(false);
     
        httpClientPolicy.setProxyServer(proxyHost);
        httpClientPolicy.setProxyServerPort(proxyPort);
        
        ProxyAuthorizationPolicy auth = new ProxyAuthorizationPolicy();
        auth.setUserName(proxyUsername);
        auth.setPassword(proxyPassword);
        
        httpConduit.setClient(httpClientPolicy);
	httpConduit.setProxyAuthorization(auth);
    }
```
and receiving this
Caused by: org.apache.cxf.transport.http.HTTPException: HTTP response '407: (POST https://webservices.cert.platform.sabre.com)

any suggestions?

**Comment by shumonsharif** [comment]
Hi @aguilardias you may want to set the proxy type using something like ...

```java
httpClientPolicy.setProxyServerType(ProxyServerType.HTTP);  // or ProxyServerType.SOCKS
```

Tested with the small tweak locally and it works fine; keep us posted if the above helps.

**Reply by aguilardias** [reply]
didnt work with setProxyServerType, how test this locally? is there a public proxy available to test?
the proxy server that im using is just available in a private cloud, i cant reach locally, so i have to deploy the application to test, im wasting a lot of time with this...

**Reply by shumonsharif** [reply]
Attached [code-with-quarkus-proxy.zip](https://github.com/quarkiverse/quarkus-cxf/files/12090085/code-with-quarkus-proxy.zip) is the project I tested with, it works without issues.  You can spin up a local proxy using:

```bash
docker run --rm -p "3128:3128/tcp" -p "1080:1080/tcp" -e "PROXY_LOGIN=ssharif" -e "PROXY_PASSWORD=password123" ghcr.io/tarampampam/3proxy:latest
```

Fire up the code-with-quarkus-proxy project and execute:

```bash
curl http://localhost:8080/add/1/2  
```

You should get a successful response.

**Reply by aguilardias** [reply]
testing with this proxy (3proxy), i've changed the code proxyserver,username,password to "whatever", and still worked, im confused if the application is using this proxy, seems dont

```
httpClientPolicy.setProxyServer("whatever");
httpClientPolicy.setProxyServerPort(1081);
httpClientPolicy.setProxyServerType(ProxyServerType.SOCKS);

ProxyAuthorizationPolicy auth = new ProxyAuthorizationPolicy();
auth.setUserName("whatever");
auth.setPassword("whatever");
```

**Reply by shumonsharif** [reply]
Aah, nice catch.  I didn't think that would be the case and assumed it was working fine ... will dig in and keep you posted.

**Reply by shumonsharif** [reply]
Looks like an issue with the SOCK proxy type.  Does it work if you switch to HTTP?

```java
    @PostConstruct
    void configurePortType() {
        final HTTPConduit httpConduit = (HTTPConduit) ClientProxy.getClient(calculatorSoap).getConduit();
        HTTPClientPolicy httpClientPolicy = new HTTPClientPolicy();
        httpClientPolicy.setProxyServer("127.0.0.1");
        httpClientPolicy.setProxyServerPort(3128);
        httpClientPolicy.setProxyServerType(ProxyServerType.HTTP);
        httpConduit.setClient(httpClientPolicy);

        ProxyAuthorizationPolicy auth = new ProxyAuthorizationPolicy();
        auth.setUserName("ssharif");
        auth.setPassword("password123");
        httpConduit.setProxyAuthorization(auth);
    }
```

**Reply by aguilardias** [reply]
Attached [code-with-quarkus-proxy-2.zip](https://github.com/quarkiverse/quarkus-cxf/files/12099646/code-with-quarkus-proxy-2.zip) is the project you sent before, with the addition of the webservice that im trying to reach (https://webservices.cert.platform.sabre.com)

with HTTP and using proxy (3proxy), CalculatorSoap worked, but my service (https://webservices.cert.platform.sabre.com/) is still receiving 
```
Caused by: org.apache.cxf.transport.http.HTTPException: HTTP response '407: (POST https://webservices.cert.platform.sabre.com) 407' when communicating with https://webservices.cert.platform.sabre.com
```
without proxy, i can reach https://webservices.cert.platform.sabre.com/, if you test, will receive
```
Caused by: jakarta.xml.ws.soap.SOAPFaultException: Authentication failed
```
this is the expected return, because im not sending Authentication credencials for security reasons.

**Reply by shumonsharif** [reply]
Okay,  so we've shown that the code works as expected.   Your proxy maintainers should be able to look into their logs and let you know why it is returning a 407.

**Reply by aguilardias** [reply, marked as answer]
2 days of work to discover that we have to set the configuration below, now its working, thanks
```
-Djdk.http.auth.tunneling.disabledSchemes=""
```
https://cxf.apache.org/docs/client-http-transport-including-ssl-support.html#ClientHTTPTransport(includingSSLsupport)-ProxyAuthentication

**Reply by shumonsharif** [reply]
Nice, glad you got to the bottom of the issue, and thanks for sharing the solution you found!
